_posts/2018-12-07-fast-starting-clojure-lambdas-using-graalvm.md

```diff
@@ -0,0 +1,80 @@
+| 1000       | JVM Java (warm)      | 5000 | 0.205      | 0.063              |
+| 1000       | JVM Clojure (cold)   | 922  | 2.942      | 0.391              |
+| 1000       | JVM Clojure (warm)   | 4600 | 0.237      | 0.133              |
+| 3000       | JVM Clojure (cold)   | 1000 | 2.348      | 0.325              |
+| 3000       | JVM Clojure (warm)   | 5000 | 0.228      | 0.134              |
+| 1000       | Python (cold)        | 1000 | 0.594      | 0.210              |
+| 1000       | Python (warm)        | 5000 | 0.358      | 0.074              |                                                                    
+                                  
+The results are quite problematic for Clojure. Using fastest available Lambda gives a start up time of 2.34 seconds. A start up time using 1000 MB Lambda is near three seconds. The long start up time makes Clojure unusable for example outgoing Slack commands which have a timout of three seconds. Slow start up is recognized in the Clojure community. Solving it may be partially accomplished with changes to Clojure itself. Fortunately new VM has been created with good results for Clojure.  
+
+## A new hope emerges - GraalVM and Lambda Custom runtime
+
+This year have given two new release which makes the situation better. First Oracle released a new [GraalVM universal virtual machine](https://www.graalvm.org/) for running applications written in JavaScript, Python, Ruby, R, JVM-based languages like Java, Scala, Kotlin, Clojure, and LLVM-based languages such as C and C++. GraalVM can ahead of time (AOT) compile JVM applications to native binaries which start very fast compared to just in time (JIT) compiled programs running in the regular JVM. Also the memory footprint is smaller in native images.
+
+Second important release was [custom AWS Lambda Runtimes](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-custom.html) which were introduced in AWS re:Invent 2018. The custom runtimes makes possible to make Lambdas with any technology which can be run on Linux - a support is now available for example Ruby, PHP and Cobol! Before that, supported were JVM, Python, Node.JS, C#, GO and PowerShell. With the custom runtime it is possible to compile Clojure as a native GraalVM binary and run it in AWS Lambda.
+
+A custom runtime API exposes API location via  environment variables. API itself contains three different REST methods for fectching invocations, posting responses and reportin errors. AWS documentation contains a useful tutorial for creating a custom runtime.
+
+## Results using GraalVM
+
+I made a simple custom runtime for running a Clojure program in a custom runtime. Compiling it was quite simple but required using a Docker container because binary must be compiled in Linux environment. Compilation times are  long compared to JVM compilation which must be made before compiling as a native image. The results are described in the next table. 
```

Comment:
```suggestion
I made a simple custom runtime for running a Clojure program in a custom runtime. Compiling it was quite simple but required using a Docker container because the binary must be compiled in Linux environment. Compilation times are long compared to JVM compilation which must be made before compiling as a native image. The results are described in the next table. 
```